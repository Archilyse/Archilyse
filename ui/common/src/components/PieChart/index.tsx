import React, { useEffect, useState } from 'react';
import { VictoryLabel, VictoryPie } from 'victory';
import C from '../../constants';
import './pieChart.scss';

const PIE_HEIGHT = 200.2;
const PIE_WIDTH = 200.3;
const DEFAULT_PIE_INNER_RADIUS = 70;

// @TODO: Use constants here
const DEFAULT_LABEL_STYLE = {
  fontSize: '20px',
  fontFamily: 'Barlow Semi Condensed',
  fontWeight: 300,
  lineHeight: '48px',
  fill: C.COLORS.SECONDARY,
};

const parseLabel = (yValue: number): string => `${parseFloat(yValue.toFixed(1))}%`;
const DEFAULT_ANIMATION_BEHAVIOUR = {};

const PieChart = ({
  data,
  animate = false,
  colorFunction = null,
  onMouseOver = (dataSource: any): any => {},
  onMouseOut = (dataSource: any): any => {},
  showLabelByDefault = false,
  innerLabel = null,
  innerRadius = null,
  labelStyle = {},
}) => {
  const [label, setLabel] = useState<string>();
  const [angle, setAngle] = useState(0);

  useEffect(() => {
    setAngle(1);
    setTimeout(() => {
      setAngle(360);
    }, 500);
  }, []);

  useEffect(() => {
    if (showLabelByDefault) {
      setLabel(parseLabel(data[0].y));
    }
  }, [data]);

  const labelText = innerLabel || label;
  if (!angle) return null; // Needed explicitly this avoid weird behaviour: https://github.com/facebook/react/issues/15281#issuecomment-480308487
  return (
    <svg viewBox={`0 0 ${PIE_HEIGHT} ${PIE_WIDTH}`}>
      <VictoryPie
        standalone={false}
        height={PIE_HEIGHT}
        width={PIE_WIDTH}
        data={data}
        endAngle={angle}
        animate={animate ? DEFAULT_ANIMATION_BEHAVIOUR : undefined}
        events={[
          {
            target: 'data',
            eventHandlers: {
              onMouseOver: () => ({
                mutation: dataSource => {
                  if (!innerLabel) setLabel(parseLabel(dataSource.datum.y));
                  // If the function has a return value
                  if (onMouseOver(dataSource)) {
                    return onMouseOver(dataSource);
                  }
                  onMouseOver(dataSource);
                },
              }),
              onMouseOut: () => ({
                mutation: dataSource => {
                  if (!showLabelByDefault && !innerLabel) {
                    setLabel('');
                  }
                  onMouseOut(dataSource);
                },
              }),
            },
          },
        ]}
        style={{
          data: {
            ...(colorFunction && { fill: colorFunction }),
            fillOpacity: 1,
            stroke: 'white',
            strokeWidth: 3,
          },
          labels: { fontSize: 8, fontFamily: 'Barlow' },
        }}
        innerRadius={innerRadius ? innerRadius : DEFAULT_PIE_INNER_RADIUS}
        labels={() => null}
      />
      {labelText && (
        <VictoryLabel
          textAnchor="middle"
          style={{ ...DEFAULT_LABEL_STYLE, ...labelStyle }}
          x={PIE_HEIGHT / 2}
          y={PIE_WIDTH / 2}
          text={labelText}
        />
      )}
    </svg>
  );
};

export default PieChart;
