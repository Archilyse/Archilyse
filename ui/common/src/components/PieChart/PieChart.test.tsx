import * as React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react';
import PieChart from './index';

afterEach(cleanup);

const MOCK_DATA = [
  { x: 'HNF', y: 50 },
  { x: 'ANF', y: 20 },
  { x: 'NNF', y: 10 },
  { x: 'FF', y: 10 },
  { x: 'VF', y: 10 },
];

const MOCK_DATA_2 = [
  { x: 'HNF', y: 50.452342423432423 },
  { x: 'ANF', y: 20 },
  { x: 'NNF', y: 10 },
  { x: 'FF', y: 10 },
  { x: 'VF', y: 10 },
];

const MOCK_COLORS = ['red', 'green', 'blue', 'yellow'];
it('renders correctly', () => {
  const { container } = render(
    <PieChart data={MOCK_DATA} showLabelByDefault={true} colorFunction={({ datum }) => MOCK_COLORS[datum._x]} />
  );
  expect(container).toMatchSnapshot();
});

it('renders correctly an y value with a lot of decimals', async () => {
  const EXPECTED_INITIAL_LABEL = new RegExp('50.5%');

  const { queryByText, getByText } = render(
    <PieChart data={MOCK_DATA_2} showLabelByDefault={true} colorFunction={({ datum }) => MOCK_COLORS[datum._x]} />
  );

  const isComponentLoaded = () => queryByText(EXPECTED_INITIAL_LABEL);
  await waitFor(() => expect(isComponentLoaded()).toBeTruthy());

  expect(getByText(EXPECTED_INITIAL_LABEL)).toBeTruthy();
});
