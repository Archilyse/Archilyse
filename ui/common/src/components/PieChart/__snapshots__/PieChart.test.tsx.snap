// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <svg
    viewBox="0 0 200.2 200.3"
  >
    <g>
      <path
        d="M4.2862637970157365e-15,-70A70,70,0,0,1,0.6108574848861728,-69.99733461449199L0.43719942846853216,-50.09809234551498A50.099999999999994,50.099999999999994,0,0,0,3.0677402318641193e-15,-50.099999999999994Z"
        role="presentation"
        shape-rendering="auto"
        style="fill: green; fill-opacity: 1; stroke: white; stroke-width: 3; padding: 10px;"
        transform="translate(100.15, 100.1)"
      />
      <path
        d="M0.6108574848861728,-69.99733461449199A70,70,0,0,1,0.8551900584672995,-69.99477587623164L0.6120717418458814,-50.09626101998864A50.099999999999994,50.099999999999994,0,0,0,0.43719942846853216,-50.09809234551498Z"
        role="presentation"
        shape-rendering="auto"
        style="fill: blue; fill-opacity: 1; stroke: white; stroke-width: 3; padding: 10px;"
        transform="translate(100.15, 100.1)"
      />
      <path
        d="M0.8551900584672995,-69.99477587623164A70,70,0,0,1,0.9773526237401746,-69.9931766806513L0.699505234991182,-50.09511645286614A50.099999999999994,50.099999999999994,0,0,0,0.6120717418458814,-50.09626101998864Z"
        role="presentation"
        shape-rendering="auto"
        style="fill: yellow; fill-opacity: 1; stroke: white; stroke-width: 3; padding: 10px;"
        transform="translate(100.15, 100.1)"
      />
      <path
        d="M0.9773526237401746,-69.9931766806513A70,70,0,0,1,1.0995122118274454,-69.99136427371624L0.7869365973222143,-50.093819287331186A50.099999999999994,50.099999999999994,0,0,0,0.699505234991182,-50.09511645286614Z"
        role="presentation"
        shape-rendering="auto"
        style="fill-opacity: 1; stroke: white; stroke-width: 3; padding: 10px;"
        transform="translate(100.15, 100.1)"
      />
      <path
        d="M1.0995122118274454,-69.99136427371624A70,70,0,0,1,1.2216684506098519,-69.98933866094738L0.8743655625079082,-50.092369527335194A50.099999999999994,50.099999999999994,0,0,0,0.7869365973222143,-50.093819287331186Z"
        role="presentation"
        shape-rendering="auto"
        style="fill-opacity: 1; stroke: white; stroke-width: 3; padding: 10px;"
        transform="translate(100.15, 100.1)"
      />
    </g>
    <text
      direction="inherit"
      dx="0"
      x="100.1"
      y="107.25"
    >
      <tspan
        dx="0"
        dy="0"
        style="font-size: 20px; font-family: Barlow Semi Condensed; font-weight: 300; line-height: 48px; fill: #898989; stroke: transparent;"
        text-anchor="middle"
        x="100.1"
      >
        50%
      </tspan>
    </text>
  </svg>
</div>
`;
